log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/instant/InstantPatternDynamicFormatter.java

```diff
@@ -711,23 +488,114 @@ public void formatTo(final StringBuilder buffer, final Instant instant) {
-            super(content, contentPrecision(content));
+        /**
+         * @param simplePattern A {@link DateTimeFormatter} pattern containing a single letter.
+         */
+        DateTimeFormatterPatternFormatterFactory(final String simplePattern) {
+            this(simplePattern, patternPrecision(simplePattern));
         }
 
         /**
-         * @param content a single-letter directive content complying (e.g., {@code H}, {@code HH}, or {@code pHH})
-         * @return the time precision of the directive
+         * @param pattern Any {@link DateTimeFormatter} pattern.
+         * @param precision The maximum interval of time over which this pattern is constant.
          */
+        DateTimeFormatterPatternFormatterFactory(final String pattern, final ChronoUnit precision) {
+            super(pattern, precision);
+        }
+
+        InstantPatternFormatter createFormatter(final Locale locale, final TimeZone timeZone) {
+            final DateTimeFormatter dateTimeFormatter =
+                    DateTimeFormatter.ofPattern(pattern, locale).withZone(timeZone.toZoneId());
+            return new AbstractFormatter(pattern, locale, timeZone, precision) {
+                @Override
+                public void formatTo(final StringBuilder buffer, final Instant instant) {
+                    final TemporalAccessor instantAccessor = toTemporalAccessor(instant);
+                    dateTimeFormatter.formatTo(instantAccessor, buffer);
+                }
+            };
+        }
+
+        @Override
         @Nullable
-        private static ChronoUnit contentPrecision(final String content) {
+        PatternFormatterFactory tryMerge(PatternFormatterFactory other, ChronoUnit thresholdPrecision) {
+            // We merge two DTF factories if they are both above or below the threshold
+            if (other instanceof DateTimeFormatterPatternFormatterFactory) {
+                final DateTimeFormatterPatternFormatterFactory otherDtf =
+                        (DateTimeFormatterPatternFormatterFactory) other;
+                if (isConstantForDurationOf(thresholdPrecision)
+                        == otherDtf.isConstantForDurationOf(thresholdPrecision)) {
+                    ChronoUnit precision = this.precision.getDuration().compareTo(otherDtf.precision.getDuration()) < 0
+                            ? this.precision
+                            : otherDtf.precision;
+                    return new DateTimeFormatterPatternFormatterFactory(this.pattern + otherDtf.pattern, precision);
+                }
+            }
+            // We merge a static pattern factory
+            if (other instanceof StaticPatternFormatterFactory) {
+                final StaticPatternFormatterFactory otherStatic = (StaticPatternFormatterFactory) other;
+                return new DateTimeFormatterPatternFormatterFactory(this.pattern + otherStatic.pattern, this.precision);
+            }
+            return super.tryMerge(other, thresholdPrecision);
```

Comment:
One less cognitive hop:
```suggestion
            return null;
```